### PR TITLE
Allows parsing debug information in the .S files

### PR DESCRIPTION
**Pull request by yurydelendik**
(Portion of #347)

The patch allows parsing of the debug information, but currently it ignores/skips it. This will allows parse files with -asm-verbose=true too.

### CODE REVIEW

File: src/compiler-support.h [hunk 1]
```diff
@@ -36,4 +36,7 @@
 # define WASM_UNREACHABLE() abort()
 #endif
 
+#define WASM_UNUSED(expr) \
```
Comment (kripken) [inline src/compiler-support.h]: this wasn't obvious to me what it does, please add a comment

File: src/s2wasm.h [hunk 2]
```diff
@@ -101,6 +101,11 @@ class S2WasmBuilder {
     abort();              \
   }
 
+  bool peek(const char *pattern) {
+    size_t size = strlen(pattern);
```
Comment (kripken) [inline src/s2wasm.h]: no need for the `size` var, can do it all on the next line

File: src/s2wasm.h [hunk 3]
```diff
@@ -449,15 +456,24 @@ class S2WasmBuilder {
   }
 
   void parseFile() {
-    assert(*s == '"');
+    if (*s != '"') {
+      // TODO: optimize, see recordFile below
+      size_t fileId = getInt();
+      skipWhitespace();
+      std::vector<char> quoted = getQuoted();
```
Comment (kripken) [inline src/s2wasm.h]: can use `auto`

File: src/s2wasm.h [hunk 4]
```diff
@@ -485,11 +501,31 @@ class S2WasmBuilder {
 
     mustMatch(":");
 
-    if (match(".Lfunc_begin")) {
+    auto recordFile = [&]() {
+      if (debug) dump("file");
+      size_t fileId = getInt();
+      skipWhitespace();
+      std::vector<char> quoted = getQuoted();
```
Comment (kripken) [inline src/s2wasm.h]: also here

File: src/s2wasm.h [hunk 5]
```diff
@@ -935,8 +981,13 @@ class S2WasmBuilder {
       } else if (match("end_block")) {
         bstack.back()->cast<Block>()->finalize();
         bstack.pop_back();
-      } else if (match(".LBB")) {
-        s = strchr(s, '\n');
+      } else if (peek(".LBB")) {
+        // FIXME legacy tests: it can be leftover from "loop" or "block", but it can be a label too
```
Comment (kripken) [inline src/s2wasm.h]: @sunfishcode: what's the status of `.LBB` and `Lfunc_end`? Looks like this PR has to introduce workarounds for them. Are they gone from `.s` now? Do we just need to update handwritten tests?

Comment (sunfishcode) [inline src/s2wasm.h]: .Lfunc_end is still in the .s file, but is emitted after the .end_func so s2wasm doesn't consider it to be part of the function body. .LBB is still present and isn't referenced from the actual code, though it probably is referenced by debug info, which is probably why this patch needs to parse it.

Comment (kripken) [inline src/s2wasm.h]: Cool, thanks @sunfishcode.

Ok, @yurydelendik, looks like the right approach to me in this pull. So aside from the comments I noted, lgtm.

**Comment by yurydelendik** [comment]
Added comment for the WASM_UNUSED macro, fixed debug.s tests output, and addressed size and auto type nits.

**Comment by kripken** [comment]
Thanks!

**Pull request merged**